### DISCUSSION TITLE
More navigation models?

### DISCUSSION THREAD

**Discussion by arkivanov** [arkivanov/Decompose#336] [Ideas]
There are Child Stack and Child Overlay high-level navigation models. There is also lower-level Generic Navigation which can be used as it is, or to create additional navigation models on top (e.g. both Child Stack and Child Overlay use it). Let's collect here use cases and feedback for additional navigation models, so most useful could be added to the library.

Please keep in mind that Decompose tries to be minimalistic with its API surface, so only reasonable and useful things should be added.

Let's do one idea per message, so we could discuss each in its own thread.

**Comment by marcguilera** [comment]
**Pager:** Similar to a stack but more like an indexed list. To simplify things that list may be static as a general use case is for onboarding slideshows, etc.

**Reply by arkivanov** [reply]
Thanks! This one looks useful. One point to consider - the amount of items can be large (e.g. 100/1000). Perhaps the API should support something like `createdSpan` argument, so only the specified amount of components around the selected index are created. It also makes sense to support changing children.

Let's gather some feedback and use cases for now.

**Reply by Shabinder** [reply]
https://github.com/arkivanov/Decompose/discussions/346
I think kinda related.

**Reply by arkivanov** [reply]
I have created a Discussion with API design for pager - #363.

**Reply by arkivanov** [reply]
An experimental version has been released in [2.0.0-beta-01](https://github.com/arkivanov/Decompose/releases/tag/2.0.0-beta-01).

**Comment by marcguilera** [comment]
**Carousel:** Very similar to a pager but I mention it separately as usually don't take the whole screen and maybe the developer we wants to be able to display 1..N items per slide depending on the screen size which impacts the NavState.

**Reply by arkivanov** [reply]
I'm wondering what are use cases? How often this is required? It feels that creating a useful API covering most of the cases would be tricky, as there are so many corner cases and different patterns. E.g., are all children always created, or should we support destroyed children? Should there be a way to deactivate (stop) children? If yes, how many children can be active at a time?

Please check the [Custom Navigation sample](https://arkivanov.github.io/Decompose/navigation/children/overview/#sample-project). It demonstrates a carousel, as well as switching to a full-screen mode.

**Reply by malliaridis** [reply]
@arkivanov Would the carousel example work too if the elements are 100 or more?

I looked into the [proposed solution for pager](https://github.com/arkivanov/Decompose/discussions/363) and thought that an additional optional parameter that allows the creation of components before index `0` and after the last index (continuing with index `0`) would also be an alternative for the carousel example in case the children exceed a specific amount.

**Reply by arkivanov** [reply]
The carousel example should work for any amount of components. You can enhance it and destroy components that are not needed, specifically for your use case.

The proposed pager API is finite. Though, I will think about making it circular. Thanks for the idea.

**Reply by malliaridis** [reply]
That's an important point I missed. So the main differences between the pager and the carousel would be the amount of children (finite/infinite) as well as the index available in the pager that is not present in the carousel?

**Reply by arkivanov** [reply]
Carousel is just an example of the Generic Navigation model. The proposed pager would also use Generic Navigation under the hood. This means that one can already create their own carousel and pager navigation, as well as many other kinds of navigation.

Currently, the idea with the pager API is to cover the specific use case of swipeable pages, with the ability to also preload side components. It's always possible to switch to Generic Navigation if something doesn't fit well.

**Reply by malliaridis** [reply]
Thanks for the clarification. I think I should try taking a more passive positioning in this topic. I feel like I am not aware of the main differences and benefits of the individual navigation models that are proposed. 😅

**Comment by marcguilera** [comment]
**Master/detail or single child**: Basically a non-nullable overlay, may be useful for bottom nav bars if the developer does not want to stack children.

**Reply by arkivanov** [reply]
I would suggest to just use Child Stack with only one child at a time, and just grab the active child in the UI. But let's gather more feedback on this.

**Reply by moffpage** [reply]
Hi! Could you provide more info on that?

**Reply by arkivanov** [reply]
Not sure who are you asking, but just in case my suggestion is to use Child Stack, and always replace the active child with another one using `replaceAll` function. In the UI, subscribe to Child Stack using `subscribeAsState` Composable function (or similar in other UI frameworks) and display the `active` child.

**Comment by pablichjenkov** [comment]
That is going to be really handy!
I'd like to see the **BottomBar** navigation pattern.

**Reply by arkivanov** [reply]
What are you missing currently for Bottom Bar navigation? It looks like Child Stack works just fine, the UI is also [pretty straightforward](https://github.com/arkivanov/Decompose/blob/master/sample/shared/compose/src/commonMain/kotlin/com/arkivanov/sample/shared/root/RootContent.kt).

**Reply by pablichjenkov** [reply]
Most of the proposed nav models  could be implemented with Children and a Child Stack. But what I had in mind was to have a **BottomBarNavigator** or **SideDrawerNavigator** component ready for use. As a user you won't need to invest time adding a BottomNavigation ui or so. Just grab the component and perhaps customize some styling.
I am basing my thoughts  on this library from react-native https://github.com/react-navigation/react-navigation, they have a nice API: https://reactnavigation.org/docs/hello-react-navigation.

**Reply by arkivanov** [reply]
Assuming we are talking about Jetpack Compose. Correct me if I'm wrong, but most of the UI code required is for composable `BottomNavigation`, see the sample - [RootContent.kt](https://github.com/arkivanov/Decompose/blob/5e9d7de05b3c4d15a9036b72a99e5927683c5742/sample/shared/compose/src/commonMain/kotlin/com/arkivanov/sample/shared/root/RootContent.kt#L48). This is the Jetpack Compose API which you have to use anyway. There is also `Children` call with a lambda switching every child, which I believe would be also required.

Are you proposing to create a unifying API for compose? Maybe you could draft an example of the API?

**Reply by pablichjenkov** [reply]
What I mean is, if is possible, having a separate module that has all these Navigators or NavModels already there.
A possible usage would be something like this:

```kotlin
@Composable
fun RootContent(rootComponent: RootComponent, modifier: Modifier = Modifier) {

   BottomNavigator(style = defaultStyle, childrenComponents = rootComponent.children) {
         Children(
            stack = childStack,
            modifier = Modifier.weight(weight = 1F),
            animation = tabAnimation(),
        ) {
            when (val child = it.instance) {
                is CountersChild -> CountersContent(component = child.component, modifier = Modifier.fillMaxSize())
                is MultiPaneChild -> MultiPaneContent(component = child.component, modifier = Modifier.fillMaxSize())
                is DynamicFeaturesChild -> DynamicFeaturesContent(component = child.component, modifier = Modifier.fillMaxSize())
                is CustomNavigationChild -> CustomNavigationContent(component = child.component, modifier.fillMaxSize())
            }
        }
   }

}
```

Or perhaps including Children in the Navigator, something like this. 

```kotlin
@Composable
fun RootContent(rootComponent: RootComponent, modifier: Modifier = Modifier) {

   BottomNavigator(style = defaultStyle, childrenComponents = rootComponent.children) {
         when (val child = it.instance) {
                is CountersChild -> CountersContent(component = child.component, modifier = Modifier.fillMaxSize())
                is MultiPaneChild -> MultiPaneContent(component = child.component, modifier = Modifier.fillMaxSize())
                is DynamicFeaturesChild -> DynamicFeaturesContent(component = child.component, modifier = Modifier.fillMaxSize())
                is CustomNavigationChild -> CustomNavigationContent(component = child.component, modifier.fillMaxSize())
            }
   }

}
```

Or lets say, nesting another Stack Navigator

```kotlin
@Composable
fun RootContent(rootComponent: RootComponent, modifier: Modifier = Modifier) {

   BottomNavigator(style = defaultStyle, childrenComponents = rootComponent.children) {
       StackNavigator {
             when (val child = it.instance) {
                is CountersChild -> CountersContent(component = child.component, modifier = Modifier.fillMaxSize())
                is MultiPaneChild -> MultiPaneContent(component = child.component, modifier = Modifier.fillMaxSize())
                is DynamicFeaturesChild -> DynamicFeaturesContent(component = child.component, modifier = Modifier.fillMaxSize())
                is CustomNavigationChild -> CustomNavigationContent(component = child.component, modifier.fillMaxSize())
            }
       }         
   }

}
```

Perhaps I am mixing concepts from the react-navigation library that don't fit properly in Decompose.

**Reply by arkivanov** [reply]
Thanks! How this API would know how to build `BottomNavigation` under the hood? If I understand the idea correctly, `BottomNavigator` would need to display the bottom bar, which should have buttons, icons, titles, etc.

**Reply by pablichjenkov** [reply]
Right, each Navigator would come with a default UI that tries to cover the component standars(lets say material 3). That is what react-navigation does, take a look into this example https://react-navigation-example.netlify.app/
Perhaps an a configuration parameter could be provided to turn on off colors and such. If a developer wants to modify it, then they would have to create their own. 
The module could have a **DefaultDecomposeDrawer** and if a developer wants something new they would need to create a **DevXDecomposeDrawer** Perhaps extending from the default. It would be a tradeoff, you write less code at the expense of using defaults.

**Reply by arkivanov** [reply]
How the implementation would know the amount of tabs, their titles and icons? Shouldn't it be also provided as some sort of configuration data class?

**Reply by pablichjenkov** [reply]
The way that library does it, for instance the TabNavigator(https://reactnavigation.org/docs/tab-based-navigation)

```JSX
<NavigationContainer>
      <Tab.Navigator
        screenOptions={({ route }) => ({
          tabBarIcon: ({ focused, color, size }) => {
            let iconName;

            if (route.name === 'Home') {
              iconName = focused
                ? 'ios-information-circle'
                : 'ios-information-circle-outline';
            } else if (route.name === 'Settings') {
              iconName = focused ? 'ios-list' : 'ios-list-outline';
            }

            // You can return any component that you like here!
            return <Ionicons name={iconName} size={size} color={color} />;
          },
          tabBarActiveTintColor: 'tomato',
          tabBarInactiveTintColor: 'gray',
        })}
      >
        <Tab.Screen name="Home" component={HomeScreen} />
        <Tab.Screen name="Settings" component={SettingsScreen} />
      </Tab.Navigator>
    </NavigationContainer>
```
<BR>
Notice the screenOption, is basically a Tab factory based on the route. Not sure how it best translate to decompose.
Perhaps something like below:


```kotlin
BottomNavigator(
     tabFactory = { childComponent ->
           when(childComponent) {
                 is CountersChild -> Tab1()
                is MultiPaneChild -> Tab2()
                is DynamicFeaturesChild -> Tab3()
                is CustomNavigationChild -> Tab4()
            }
     }, 
     childrenComponents = rootComponent.children
) {
          when (val child = it.instance) {
                is CountersChild -> CountersContent(component = child.component, modifier = Modifier.fillMaxSize())
                is MultiPaneChild -> MultiPaneContent(component = child.component, modifier = Modifier.fillMaxSize())
                is DynamicFeaturesChild -> DynamicFeaturesContent(component = child.component, modifier = Modifier.fillMaxSize())
                is CustomNavigationChild -> CustomNavigationContent(component = child.component, modifier.fillMaxSize())
            }
}
```

**Reply by arkivanov** [reply]
Thanks for the clarification!

**Reply by malliaridis** [reply]
If I am getting it right this would mix UI components with navigation logic and make it harder to maintain over time, especially if Decompose focuses on multiplatform support. I think the current `ChildStack` covers this use-case well without direct dependencies to UI libraries. But what might be beneficial in regards to this idea would be to add simplified abstractions of `ChildStack` implementations for different scenarios, naming them accordingly.

Just for my understanding, would `BottomNavigator` in the example above need to be a `Composable` function if Compose would be used?

**Comment by pablichjenkov** [comment]
**NavigationDrawer**: Navigation drawer when the screen is small and left/right side panel when the screen is big.

**Reply by arkivanov** [reply]
Interesting! At the first glance, I would just try using Child Overlay. The UI could signal the screen type to the component, and the component would control whether the overlay is active or not. Just a rough idea. Let's gather more feedback.

**Reply by malliaridis** [reply]
The [M3 `NavigationDrawer`](https://m3.material.io/components/navigation-drawer/guidelines#159bffe1-2fea-4d47-9e50-e2cfa0ec37ef) has two different types, standard and modal. And on larger screens one might consider using a [`NavigationRail`](https://m3.material.io/components/navigation-rail/overview) (additional UI component) for opening a `NavigationDrawer` of type modal on larger screens instead of a permanent side-panel / `NavigationDrawer` of type standard.

How would the different scenarios be taken into account in case Decompose would provide a `NavigationDrawer`?

**Reply by arkivanov** [reply]
A related discussion that might be useful: #430
